File: go-backend-interview-prep/01_go_basics/08_functions_intro/main.py
# Python equivalent of Functions Intro (compare with main.go)

def greet(name: str) -> str:
    return "hello, " + name


def add(a: int, b: int) -> int:
    return a + b


def log_message(msg: str) -> None:
    print("[LOG]", msg)


def describe(name: str, age: int) -> str:
    return f"{name} is {age} years old"


def apply(value: int, fn) -> int:
    """Takes a function as a parameter (first-class functions)."""
    return fn(value)


def no_change(x: int) -> None:
    x = 999  # rebinds local name; original int is immutable


def main():
    # --- Calling functions ---
    print(greet("world"))
    print("3 + 7 =", add(3, 7))
    log_message("server started")
    print(describe("Alice", 30))

    # --- Functions as values ---
    double = lambda x: x * 2
    print("double(5):", double(5))

    # --- Passing a function as an argument ---
    result = apply(10, double)
    print("apply(10, double):", result)

    # --- Inline anonymous function (lambda) ---
    print("inline:", apply(4, lambda x: x * x))

    # --- Pass-by-value for immutables ---
    original = 42
    no_change(original)
    print("after no_change, original:", original)  # still 42


if __name__ == "__main__":
    main()

File: go-backend-interview-prep/01_go_basics/08_functions_intro/main.go
package main

import "fmt"

// --- Basic function ---
func greet(name string) string {
	return "hello, " + name
}

// --- Same-type parameter shorthand ---
func add(a, b int) int {
	return a + b
}

// --- No return value ---
func logMessage(msg string) {
	fmt.Println("[LOG]", msg)
}

// --- Multiple types ---
func describe(name string, age int) string {
	return fmt.Sprintf("%s is %d years old", name, age)
}

func main() {
	// --- Calling functions ---
	fmt.Println(greet("world"))
	fmt.Println("3 + 7 =", add(3, 7))
	logMessage("server started")
	fmt.Println(describe("Alice", 30))

	// --- Functions as values ---
	double := func(x int) int {
		return x * 2
	}
	fmt.Println("double(5):", double(5))

	// --- Passing a function as an argument ---
	result := apply(10, double)
	fmt.Println("apply(10, double):", result)

	// --- Inline anonymous function ---
	fmt.Println("inline:", apply(4, func(x int) int {
		return x * x
	}))

	// --- Pass-by-value demonstration ---
	original := 42
	noChange(original)
	fmt.Println("after noChange, original:", original) // still 42
}

// --- Function that takes a function as a parameter ---
func apply(value int, fn func(int) int) int {
	return fn(value)
}

// --- Demonstrates pass-by-value ---
func noChange(x int) {
	x = 999 // modifies the copy, not the original
}
